compiler/qsc_data_structures/src/namespaces.rs

```diff
@@ -0,0 +1,304 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#[cfg(test)]
+mod tests;
+
+use rustc_hash::FxHashMap;
+use std::{cell::RefCell, fmt::Display, iter::Peekable, ops::Deref, rc::Rc};
+
+pub const PRELUDE: [[&str; 3]; 4] = [
+    ["Microsoft", "Quantum", "Canon"],
+    ["Microsoft", "Quantum", "Core"],
+    ["Microsoft", "Quantum", "Intrinsic"],
+    ["Microsoft", "Quantum", "Measurement"],
+];
+
+/// An ID that corresponds to a namespace in the global scope.
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq, Default)]
+pub struct NamespaceId(usize);
+impl NamespaceId {
+    /// Create a new namespace ID.
+    #[must_use]
+    pub fn new(value: usize) -> Self {
+        Self(value)
+    }
+}
+
+impl From<usize> for NamespaceId {
+    fn from(value: usize) -> Self {
+        Self::new(value)
+    }
+}
+
+impl From<NamespaceId> for usize {
+    fn from(value: NamespaceId) -> Self {
+        value.0
+    }
+}
+
+impl From<&NamespaceId> for usize {
+    fn from(value: &NamespaceId) -> Self {
+        value.0
+    }
+}
+
+impl Deref for NamespaceId {
+    type Target = usize;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl Display for NamespaceId {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "Namespace {}", self.0)
+    }
+}
+
+/// A reference counted cell that supports interior mutability for namespace tree nodes.
+/// Interior mutability is required to update the tree when inserting new data structures.
+type NamespaceTreeCell = Rc<RefCell<NamespaceTreeNode>>;
+
+/// An entry in the memoization table for namespace ID lookups.
+type MemoEntry = (Vec<Rc<str>>, NamespaceTreeCell);
+
+/// The root of the data structure that represents the namespaces in a program.
+/// The tree is a trie (prefix tree) where each node is a namespace and the children are the sub-namespaces.
+/// For example, the namespace `Microsoft.Quantum.Canon` would be represented as a traversal from the root node:
+/// ```
+/// root
+/// └ Microsoft
+///   └ Quantum
+///     └ Canon
+/// ```
+/// This data structure is optimized for looking up namespace IDs by a given name. Looking up a namespace name by ID is
+/// less efficient, as it performs a breadth-first search. Because of this inefficiency, the results of this lookup are memoized.
+/// [`NamespaceTreeNode`]s are all stored in [`NamespaceTreeCell`]s, which are reference counted and support interior mutability for namespace
+/// insertions and clone-free lookups.
+#[derive(Debug, Clone)]
+pub struct NamespaceTreeRoot {
+    assigner: usize,
+    tree: NamespaceTreeCell,
+    memo: RefCell<FxHashMap<NamespaceId, MemoEntry>>,
```

Comment:
Not an objection, but just curious: Why is this inside a `RefCell`? My guess is "to avoid updating a bunch of existing places to be `&mut` references" - is that right?